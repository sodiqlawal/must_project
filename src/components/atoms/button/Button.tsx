import clx from 'classnames';
import { FC, ButtonHTMLAttributes, PropsWithChildren } from 'react';
import styles from './Button.module.scss';

export type ButtonProps = {
  className: string;
  isPrimary: boolean;
  isOutline: boolean;
  isTransparent: boolean;
  w: string;
  h: string;
};

const Button: FC<
  PropsWithChildren<
    ButtonHTMLAttributes<HTMLButtonElement> & Partial<ButtonProps>
  >
> = ({
  children,
  className,
  isPrimary,
  isOutline,
  isTransparent,
  w,
  h,
  ...props
}) => {
  return (
    <button
      type={props.type || 'button'}
      className={clx(styles.btn, className, {
        [styles.primary]: isPrimary,
        [styles.outline]: isOutline,
        [styles.transparent]: isTransparent,
        [styles.disabled]: props.disabled,
      })}
      disabled={props.disabled}
      style={{
        width: w || 'inherit',
        height: h || 'inherit',
      }}
      {...props}
    >
      {children}
    </button>
  );
};
export default Button;
