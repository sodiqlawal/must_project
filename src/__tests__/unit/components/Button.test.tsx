import { render, screen, fireEvent } from "@testing-library/react";
import Button from "@/components/atoms/button/Button";
import { ReactElement } from "react";

const defaultProps: {
    children: any;
    onClick: () => void;
    type: "submit" | "button";
    disabled: boolean;
} = {
    children: <div>Submit</div>,
    onClick: jest.fn(),
    type: "submit",
    disabled: false,
};

const setupButton = (props = defaultProps) => render(<Button {...props} />);

describe("The <Button /> component", () => {

    it("should match snapshot", () => {
        const { asFragment } = setupButton();

        expect(asFragment()).toMatchSnapshot()
    });

    it("should render the button and content", () => {
        setupButton();

        const button = screen.getByRole("button", { name: /Submit/i });

        expect(button).toBeInTheDocument();
    });

    it("should call the click handler", () => {
        setupButton();

        const button = screen.getByRole("button", { name: /Submit/i });

        fireEvent.click(button);

        expect(defaultProps.onClick).toHaveBeenCalled();
    });

    it("should disable button if disabled is true", () => {
        defaultProps.disabled = true;
        setupButton();

        const button = screen.getByRole("button", { name: /Submit/i });

        expect(button).toBeDisabled();
    });

});
