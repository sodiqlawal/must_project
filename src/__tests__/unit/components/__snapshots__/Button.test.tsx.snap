// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Button /> component should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
    style="width: inherit; height: inherit;"
    type="submit"
  >
    <div>
      Submit
    </div>
  </button>
</DocumentFragment>
`;
